### PR TITLE
generic_factory: ordered init. of static globals

### PR DESCRIPTION
**Pull request by andrei8l**
#### Summary
None

#### Purpose of change
Some global variables depend on other global variables but their order of initialization is not well defined
Fixes: https://github.com/CleverRaven/Cataclysm-DDA/issues/56813

#### Describe the solution
Initialize dependent static variables in the appropriate order

#### Describe alternatives you've considered
This seems to be the simplest, least invasive solution

#### Testing
gcc build with `-O3 -flto` (and `-fsanitize=address,undefined` if you're willing to wait a few hours) doesn't crash

LTO workflow should not segfault

#### Additional context
BrettDong figured it out first

### CODE REVIEW

File: src/ammo_effect.cpp [hunk 1]
```diff
@@ -6,53 +6,52 @@
 #include "generic_factory.h"
 #include "json.h"
 
-namespace
+generic_factory<ammo_effect> *get_all_ammo_effects()
```
Comment (Qrox) [inline src/ammo_effect.cpp:9]:
```suggestion
generic_factory<ammo_effect> &get_all_ammo_effects()
```
Why not return a reference?

Comment (andrei8l) [reply src/ammo_effect.cpp:9]: What difference does that make here? akrieger suggested the same thing so I must be missing something.

Comment (Qrox) [reply src/ammo_effect.cpp:9]: It is better to return a reference if the return value can never be null, so the caller does not have to worry about dereferencing a null pointer. It also allows you to use the return value in the same way as a dereferenced value, so you don't have to replace `.` with `->`, use `*` to dereference it, or use `&` to get the address when returning.

Comment (andrei8l) [reply src/ammo_effect.cpp]: Alright, thanks.

**Pull request merged**